File: src/main/java/me/sd_master92/customvoting/database/PlayerTable.java
package me.sd_master92.customvoting.database;

import me.sd_master92.customfile.database.CustomColumn;
import me.sd_master92.customfile.database.CustomDatabase;
import me.sd_master92.customfile.database.CustomTable;
import me.sd_master92.customvoting.Main;
import me.sd_master92.customvoting.extensions.CustomPlaceholders;
import org.bukkit.scheduler.BukkitRunnable;

import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

public class PlayerTable
{
    private final Main plugin;
    private final CustomTable players;

    public PlayerTable(Main plugin, CustomDatabase database)
    {
        this.plugin = plugin;
        players = database.getTable("players");
        if (!players.exists())
        {
            if (!players.create("uuid", CustomColumn.DataType.VARCHAR_PRIMARY))
            {
                plugin.print("| could not create table 'players'");
                plugin.print("|");
                plugin.print("|___database disabled");
                return;
            } else
            {
                players.getColumn("name").create(CustomColumn.DataType.VARCHAR);
                players.getColumn("votes").create(CustomColumn.DataType.INT);
                players.getColumn("last").create(CustomColumn.DataType.LONG);
                players.getColumn("queue").create(CustomColumn.DataType.INT);
                plugin.print("| successfully created table 'players'");
            }
        } else
        {
            plugin.print("| successfully located table 'players'");
        }
        plugin.print("|");
        plugin.print("|___successfully connected to database");
    }

    public static List<PlayerData> getTopVoters(Main plugin)
    {
        List<PlayerData> topVoters = new ArrayList<>();
        try
        {
            ResultSet all = plugin.getPlayerTable().getAll();
            while (all.next())
            {
                topVoters.add(new PlayerData(all.getString("uuid"), all.getString("name"), all.getInt("votes"),
                        all.getLong("last"),
                        all.getInt("queue")));
            }
        } catch (Exception e)
        {
            return topVoters;
        }

        topVoters.sort((x, y) ->
        {
            int compare = Integer.compare(y.getVotes(), x.getVotes());
            if (compare == 0)
            {
                compare = Long.compare(x.getLast(), y.getLast());
            }
            return compare;
        });

        if (plugin.usePlaceholders())
        {
            new BukkitRunnable()
            {
                @Override
                public void run()
                {
                    for (int i = 0; i < topVoters.size(); i++)
                    {
                        CustomPlaceholders.setPlayerVotes(i + 1);
                    }
                }
            }.runTask(plugin);
        }

        return topVoters;
    }

    public static PlayerData getTopVoter(Main plugin, int n)
    {
        n--;
        List<PlayerData> topVoters = getTopVoters(plugin);
        if (n >= 0 && n < topVoters.size())
        {
            return topVoters.get(n);
        }
        return null;
    }

    public CustomTable getTable()
    {
        return players;
    }

    public boolean addPlayer(String uuid)
    {
        return players.insertData(new String[]{"uuid", "votes", "last", "queue"}, new String[]{"\"" + uuid + "\"", "0",
                "0", "0"});
    }

    public ResultSet getAll()
    {
        return players.getAll();
    }

    public String getUuid(String name)
    {
        ResultSet result = players.getData("name=\"" + name + "\"");
        try
        {
            if (result.first())
            {
                return result.getString("uuid");
            }
        } catch (Exception e)
        {
            plugin.error("Could not retrieve uuid of " + name + " from database");
        }
        return "Unknown";
    }

    public String getName(String uuid)
    {
        ResultSet result = players.getData("uuid=\"" + uuid + "\"");
        try
        {
            if (result.first())
            {
                return result.getString("name");
            }
        } catch (Exception e)
        {
            plugin.error("Could not retrieve name of " + uuid + " from database");
        }
        return "Unknown";
    }

    public boolean setName(String uuid, String name)
    {
        return players.updateData("uuid=\"" + uuid + "\"", "name=\"" + name + "\"");
    }

    public int getVotes(String uuid)
    {
        ResultSet result = players.getData("uuid=\"" + uuid + "\"");
        try
        {
            if (result.first())
            {
                return result.getInt("votes");
            } else
            {
                addPlayer(uuid);
            }
        } catch (Exception e)
        {
            plugin.error("Could not retrieve votes of " + uuid + " from database");
        }
        return 0;
    }

    public boolean setVotes(String uuid, int votes)
    {
        return players.updateData("uuid=\"" + uuid + "\"", "votes=" + votes);
    }

    public long getLast(String uuid)
    {
        ResultSet result = players.getData("uuid=\"" + uuid + "\"");
        try
        {
            if (result.first())
            {
                return result.getLong("last");
            } else
            {
                addPlayer(uuid);
            }
        } catch (Exception e)
        {
            plugin.error("Could not retrieve last timestamp of " + uuid + " from database");
        }
        return 0;
    }

    public boolean setLast(String uuid)
    {
        return players.updateData("uuid=\"" + uuid + "\"", "last=" + System.currentTimeMillis());
    }

    public int getQueue(String uuid)
    {
        ResultSet result = players.getData("uuid=\"" + uuid + "\"");
        try
        {
            if (result.first())
            {
                return result.getInt("queue");
            } else
            {
                addPlayer(uuid);
            }
        } catch (Exception e)
        {
            plugin.error("Could not retrieve queue of " + uuid + " from database");
        }
        return 0;
    }

    public boolean setQueue(String uuid, int queue)
    {
        return players.updateData("uuid=\"" + uuid + "\"", "queue=" + queue);
    }
}

File: src/main/kotlin/me/sd_master92/customvoting/database/PlayerTable.kt
package me.sd_master92.customvoting.database

import com.github.shynixn.mccoroutine.bukkit.launch
import kotlinx.coroutines.runBlocking
import me.sd_master92.customvoting.CV
import me.sd_master92.customvoting.constants.enumerations.PMessage
import me.sd_master92.customvoting.constants.enumerations.Setting
import me.sd_master92.customvoting.constants.interfaces.TopVoter
import me.sd_master92.customvoting.constants.interfaces.Voter
import me.sd_master92.customvoting.constants.models.VoteHistory
import me.sd_master92.customvoting.constants.models.VoteSiteUUID
import me.sd_master92.customvoting.dayDifference
import org.bukkit.entity.Player
import java.util.*

class PlayerTable(private val plugin: CV, val uuid: UUID) : Voter
{
    private val players: PlayerDatabase? = plugin.playerDatabase

    private constructor(plugin: CV, player: Player) : this(plugin, player.uniqueId)
    {
        plugin.launch {
            players?.setName(player.uniqueId, player.name)
        }
    }

    private suspend fun register()
    {
        players?.getVotes(uuid)
        ALL[uuid] = this
    }

    override suspend fun getUuid(): UUID
    {
        return uuid
    }

    override suspend fun getName(): String
    {
        return players?.getName(uuid) ?: PMessage.PLAYER_NAME_UNKNOWN.toString()
    }

    override suspend fun setName(name: String): Boolean
    {
        return players?.setName(uuid, name) ?: false
    }

    override suspend fun getVotes(): Int
    {
        return players?.getVotes(uuid) ?: 0
    }

    override suspend fun getVotesMonthly(): Int
    {
        return players?.getMonthlyVotes(uuid) ?: 0
    }

    override suspend fun getVotesWeekly(): Int
    {
        return players?.getWeeklyVotes(uuid) ?: 0
    }

    override suspend fun getVotesDaily(): Int
    {
        return players?.getDailyVotes(uuid) ?: 0
    }

    override suspend fun getPower(): Boolean
    {
        return players?.getPower(uuid) == true
    }

    override suspend fun getLast(): Long
    {
        return getHistory().maxByOrNull { it.time }?.time ?: 0
    }

    override suspend fun getHistory(): List<VoteHistory>
    {
        return players?.getHistory(uuid) ?: listOf()
    }

    override suspend fun getStreakDaily(): Int
    {
        return players?.getStreak(uuid) ?: 0
    }

    override suspend fun setStreakDaily(n: Int): Boolean
    {
        return players?.setStreak(uuid, n) ?: false
    }

    override suspend fun setVotes(n: Int, update: Boolean)
    {
        plugin.launch {
            players?.setVotes(uuid, n)
            players?.setMonthlyVotes(uuid, 0)
            players?.setWeeklyVotes(uuid, 0)
            players?.setDailyVotes(uuid, 0)

            if (update)
            {
                Voter.getTopVoters(plugin, true)
            }
        }
    }

    override suspend fun addVote(site: VoteSiteUUID, queued: Boolean): Boolean
    {
        val votesBefore = getVotes()
        addStreak()
        addHistory(site, queued)
        players?.setVotes(uuid, getVotes() + 1)
        players?.setMonthlyVotes(uuid, getVotesMonthly() + 1)
        players?.setWeeklyVotes(uuid, getVotesWeekly() + 1)
        players?.setDailyVotes(uuid, getVotesDaily() + 1)

        Voter.getTopVoters(plugin, true)

        return votesBefore < getVotes()
    }

    override suspend fun setPower(power: Boolean): Boolean
    {
        return players?.setPower(uuid, power) ?: false
    }

    override suspend fun clearQueue(): Boolean
    {
        return players?.clearQueue(uuid) ?: false
    }

    override suspend fun addHistory(site: VoteSiteUUID, queued: Boolean): Boolean
    {
        return players?.addHistory(uuid, site, queued) ?: false
    }

    override suspend fun addStreak(): Boolean
    {
        val diff = getLast().dayDifference()
        if (!plugin.config.getBoolean(Setting.VOTE_STREAK_CONSECUTIVE.path) || diff == 1 || getVotes() == 0)
        {
            return players?.setStreak(uuid, getStreakDaily() + 1) ?: false
        } else if (diff > 1)
        {
            clearStreak()
        }
        return false
    }

    override suspend fun clearStreak(): Boolean
    {
        return players?.setStreak(uuid, 0) ?: false
    }

    override suspend fun clearMonthlyVotes()
    {
        players?.setMonthlyVotes(uuid, 0)
    }

    override suspend fun clearWeeklyVotes()
    {
        players?.setWeeklyVotes(uuid, 0) ?: false
    }

    override suspend fun clearDailyVotes()
    {
        players?.setDailyVotes(uuid, 0) ?: false
    }

    companion object : TopVoter
    {
        private var ALL: MutableMap<UUID, PlayerTable> = HashMap()

        fun init(plugin: CV)
        {
            plugin.launch {
                val result = plugin.playerDatabase?.playersTable?.getAllAsync()
                if (result != null)
                {
                    while (result.next())
                    {
                        try
                        {
                            PlayerTable(plugin, UUID.fromString(result.getString(PlayerTableColumn.UUID.columnName)))
                        } catch (_: Exception)
                        {
                        }
                    }
                }
            }
        }

        override fun getAll(): MutableList<Voter>
        {
            return ALL.values.toMutableList()
        }

        suspend fun get(plugin: CV, player: Player): PlayerTable
        {
            return if (plugin.config.getBoolean(Setting.UUID_STORAGE.path)) getByUuid(plugin, player) else getByName(
                plugin,
                player
            )
        }

        private fun getByUuid(plugin: CV, player: Player): PlayerTable
        {
            return ALL.getOrElse(player.uniqueId) {
                val voter = PlayerTable(plugin, player)
                ALL[player.uniqueId] = voter
                return voter
            }
        }

        private suspend fun getByName(plugin: CV, player: Player): PlayerTable
        {
            var voter = ALL.values.firstOrNull { file -> file.getName() == player.name }
            if (voter == null)
            {
                voter = PlayerTable(plugin, player)
                ALL[player.uniqueId] = voter
            }
            return voter
        }
    }

    init
    {
        runBlocking {
            register()
        }
    }
}